innovation.game.php

```diff
@@ -10878,7 +10908,86 @@ function stInteractionStep() {
                 'with_icon' => 1 /* tower */
             );
             break;
+
+        // id 155, Artifacts age 5: Boerhavve Silver Microscope
+        case "155N1A":
+            //Return the lowest card in your hand 
+            $min_hand_age = self::getMinAgeInHand($player_id);
+            $options = array(
+                'player_id' => $player_id,
+                'n' => 1,
+                'can_pass' => false,
+                
+                'owner_from' => $player_id,
+                'location_from' => 'hand',
+                'owner_to' => 0,
+                'location_to' => 'deck',
+                'auxiliary_value' => $min_hand_age,
+                'age' => $min_hand_age
+            );
+
+            break;
             
+        case "155N1B":
+            // and the lowest top card on your board.
+            $all_cards = self::getTopCardsOnBoard($player_id);
+            $ages = array();
+            foreach($all_cards as $card)
+            {
+                $ages[] = $card['age'];
+            }
+            
+            $options = array(
+                'player_id' => $player_id,
+                'n' => 1,
+                'can_pass' => false,
+                
+                'owner_from' => $player_id,
+                'location_from' => 'board',
+                'owner_to' => 0,
+                'location_to' => 'deck',
+                
+                'age' => min($ages)
+            );
+            break;
+            
+        // id 156, Artifacts age 5: Principia
+        case "156N1A":
+            $ages_on_top = array();
+            for($color = 1; $color < 5; $color++) {
+                $topcard = self::getTopCardOnBoard($player_id, $color);
+                if ($topcard !== null){ // skip empty piles
+                    $ages_on_top[] = $topcard['age']; // Record all non-blue ages
+                }
+            }
+            self::setGameStateValueFromArray('auxiliary_value', $ages_on_top);
```

Comment:
The array you pass into setGameStateValueFromArray cannot have duplicate values or it won't work correctly.

I haven't tested the following code, but I believe something like this would work (other cards do something similar):

$ages_on_top = array();
self::setGameStateValueAsArray('auxiliary_value', $ages_on_top);
for($color = 1; $color < 5; $color++) { // non-blue
    $top_card = self::getTopCardOnBoard($player_id, $color);
    if ($top_card !== null) {
        $ages_on_top[] = $top_card['age'];
        if (!in_array($top_card['age'], $ages_on_top)) {
            $ages_on_top[] = $top_card['age'];
            self::setGameStateValueFromArray('auxiliary_value', $ages_on_top);
        }
    }
}

Reply:
hmm...this array needs to have duplicates in it.  (i.e. you can return more than one card of the same age)

This limitation of arrays is really annoying...heh.

Reply:
After looking at the code for the array encoding, I have an idea for a different type of encoding which will make arrays of "ages" much easier.  The maximum value of the integers in this framework is insanely small though (0x8000000).  Why not a full 32 bit number?  I think adding additional "auxiliary_value" fields will be needed moving forward to augment the size of the arrays.

Reply:
Ok, this is what I have:

   // This function encodes multiple integers into a single array
    // which can be decoded by getGameStateBase16Array.
    // Note that due to the maximum data value of 0x8000000 that only
    // 5 elements can be encoded in this manner.
    // This function is ideal for encoding multiple age values into an array.
    function setGameStateBase16Array($key, $array) {
        if (count($array) > 5) {
            // Too many elements to encode.  Throw an error.
        }
        $encoded_value = 0;
        foreach($array as $value) {
            $encoded_value = $encoded_value * 16 + $value;
        }
        $encoded_value = $encoded_value * 6 + count($array);
        self::setGameStateValue($key, $encoded_value);
    }
    
    // This is the decoding partner of setGameStateBase16Array
    function getGameStateBase16Array($key) {
        $return_array = array();
        $encoded_value = self::getGameStateValue($key);
        $count = ($encoded_value % 6); // count is first
        $encoded_value /= 6;
        for ($i = 0; $i < $count; $i++) {
            $return_array[] = $encoded_value % 16;
            $encoded_value /= 16;
        }
        return $return_array;
    }
    
It can have a maximum size of 5 and a minimum of 0.  I tested it out with Principia and it works pretty well.

Reply:
Great! Let me know when you've pushed your changes and want me to take a look at this PR.